server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipServiceTests.java

```diff
@@ -0,0 +1,291 @@
+
+        applyClusterState("add new nodes", this::withNewNodes);
+        applyClusterState("add desired nodes", this::desiredNodesWithAllClusterNodes);
+
+        final var desiredNodes = DesiredNodes.latestFromClusterState(clusterService.state());
+        final var clusterState = clusterService.state();
+        final var leavingNode = randomValueOtherThan(clusterState.nodes().getMasterNode(), () -> randomFrom(clusterState.nodes()));
+        final var discoveryNodes = DiscoveryNodes.builder(clusterState.nodes()).remove(leavingNode);
+        final var leavingDesiredNode = desiredNodes.find(leavingNode.getExternalId());
+
+        applyClusterState("remove some nodes", state -> ClusterState.builder(state).nodes(discoveryNodes).build());
+
+        assertThat(tracker.isMember(leavingDesiredNode), is(equalTo(false)));
+
+        applyClusterState(
+            "Add node back",
+            state -> ClusterState.builder(state)
+                .nodes(DiscoveryNodes.builder(state.nodes()).add(newNode(leavingDesiredNode.externalId())))
+                .build()
+        );
+
+        assertThat(tracker.isMember(leavingDesiredNode), is(equalTo(true)));
+    }
+
+    public void testNodeLeavesAndJoinsWithNewHistoryId() {
+        final var tracker = DesiredNodesMembershipService.create(clusterService);
+
+        applyClusterState("add new nodes", this::withNewNodes);
+        applyClusterState("add desired nodes", this::desiredNodesWithAllClusterNodes);
+
+        final var desiredNodes = DesiredNodes.latestFromClusterState(clusterService.state());
+        final var clusterState = clusterService.state();
+        final var leavingNode = randomValueOtherThan(clusterState.nodes().getMasterNode(), () -> randomFrom(clusterState.nodes()));
+        final var discoveryNodes = DiscoveryNodes.builder(clusterState.nodes()).remove(leavingNode);
+        final var leavingDesiredNode = desiredNodes.find(leavingNode.getExternalId());
+
+        applyClusterState("remove some nodes", state -> ClusterState.builder(state).nodes(discoveryNodes).build());
+
+        assertThat(tracker.isMember(leavingDesiredNode), is(equalTo(false)));
+
+        final var desiredNodesWithNewHistoryId = new DesiredNodes(UUIDs.randomBase64UUID(), 1, desiredNodes.nodes());
+
+        applyClusterState(
+            "Add node back and a new desired nodes",
+            state -> ClusterState.builder(state)
+                .nodes(DiscoveryNodes.builder(clusterState.nodes()))
+                .metadata(
+                    Metadata.builder(state.metadata())
+                        .putCustom(DesiredNodesMetadata.TYPE, new DesiredNodesMetadata(desiredNodesWithNewHistoryId))
+                        .build()
+                )
+                .build()
+        );
+
+        for (DesiredNode desiredNode : desiredNodesWithNewHistoryId) {
+            assertThat(tracker.isMember(desiredNode), is(equalTo(true)));
+        }
+    }
+
+    public void testMasterDemotionClearsMembers() {
```

Comment:
I think we need a test demonstrating that a promoted master has the right set of members?